pkg/driver/node_server.go

```diff
@@ -1028,152 +1015,125 @@ func (driver *Driver) nodePublishEphemeralVolume(
-	sizeInBytes := defaultVolumeSize
-
-	// Get the volume size from the volume context if specified
-	sizeStr := volumeContext[sizeKey]
-	if sizeStr != "" {
-		volSize := resource.MustParse(sizeStr)
-		sizeInBytes = volSize.Value()
-		log.Tracef("Ephemeral volume %s requested with size %s (%v)", volumeName, sizeStr, sizeInBytes)
+	// Get Secret
+	secrets, secretRef, err := driver.getEphemeralVolumeSecret(volumeHandle, volumeContext)
+	if err != nil {
+		return status.Error(codes.InvalidArgument,
+			fmt.Sprintf("Failed to node publish ephemeral volume %s, %s", volumeHandle, err.Error()))
 	}
 
-	// Construct volume capabitilities to pass to createVolume()
-	volCapabilities := []*csi.VolumeCapability{
+	// Create and Publish inline ephemeral volume
+	log.Infof("Provisioning ephemeral volume %s with name %s for pod %s", volumeHandle, volumeName, podUID)
+	err = driver.createPublishEphemeralVolume(
+		volumeHandle,
+		volumeName,
+		stagingTargetPath,
+		targetPath,
 		volumeCapability,
-	}
-
-	// Check if the requested volume already exists on the backend.
-	// If yes, then we need to force re-create it (destroy and then create).
-	{
-		volume, err := driver.GetVolumeByName(volumeName, secrets)
-		if err != nil {
-			log.Errorf("Error while trying to check if ephemeral volume %s already exists, err: %s", volumeName, err.Error())
-			return status.Error(codes.Internal,
-				fmt.Sprintf("NodePublish of ephemeral volume %s failed, %s", volumeName, err.Error()))
+		secrets,
+		secretRef,
+		readOnly,
+		publishContext,
+		volumeContext,
+	)
+	if err != nil {
+		log.Errorf("Failed to node publish ephemeral volume %s, attempting to rollback, err: %s", volumeName, err.Error())
+		if status.Code(err) == codes.AlreadyExists {
+			log.Errorf("Failed to rollback ephemeral volume %s, err: %s", volumeName, err.Error())
+			return err
 		}
-		// If volume found, then force destroy it.
-		if volume != nil {
-			log.Infof("Volume %s already exists, attempting to force destroy", volumeName)
-			err = driver.deleteVolume(volume.ID, secrets, true)
-			if err != nil {
-				return status.Error(codes.Internal,
-					fmt.Sprintf("NodePublish of ephemeral volume %s failed, unable to destroy pre-existing volume with id %s, %s",
-						volumeName, volume.ID, err.Error()))
-			}
-			log.Infof("Successfully destroyed the pre-existing ephemeral volume %s", volumeName)
+		// Attempt rollback
+		rbErr := driver.retryRollbackEphemeralVolume(
+			volumeHandle,
+			volumeName,
+			secrets,
+			stagingTargetPath,
+			targetPath,
+		)
+		if rbErr != nil {
+			log.Errorf("Failed to rollback ephemeral volume %s, err: %s", volumeName, rbErr.Error())
+			// Returning both original error + rollback error
+			return status.Error(codes.Internal,
+				fmt.Sprintf("%s, %s", err.Error(), rbErr.Error()))
 		}
+		log.Infof("Rollback successful for ephemeral volume %s", volumeName)
+		return err
 	}
+	return nil
+}
+
+func (driver *Driver) createPublishEphemeralVolume(
```

Comment:
is there a reason to tie these two together? can create and publish volume be still kept separate?

Reply:
`createPublishEphemeralVolume` is just another separate wrapper function which internally calls create(), controlPublish(), nodeStage() and nodePublish() separately. There are grouped together so that we can rollback if any failure occurs within those intermediate steps.

Reply:
got it. thanks.